File: leet-code/fourSumCount.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @param {number[]} C
 * @param {number[]} D
 * @return {number}
 */
var fourSumCount = function (A, B, C, D) {
  const countMap = new Map()
  const len = A.length
  let res = 0
  for (let i = 0; i < len; i++) {
    for (let j = 0; j < len; j++) {
      const val = A[i] + B[j]
      if (countMap.has(val)) {
        countMap.set(val, countMap.get(val) + 1)
      } else {
        countMap.set(val, 1)
      }
    }
  }
  for (let i = 0; i < len; i++) {
    for (let j = 0; j < len; j++) {
      if (countMap.has(-C[i] - D[j])) {
        res += countMap.get(-C[i] - D[j])
      }
    }
  }
  return res
}

File: leet-code/fourSumCount.go
package main

func fourSumCount(A []int, B []int, C []int, D []int) int {
	m := len(A)
	res := 0
	countMap := make(map[int]int)
	for i := 0; i < m; i++ {
		for j := 0; j < m; j++ {
			val := A[i] + B[j]
			if _, ok := countMap[val]; ok {
				countMap[val]++
			} else {
				countMap[val] = 1
			}
		}
	}

	for i := 0; i < m; i++ {
		for j := 0; j < m; j++ {
			val := -C[i] - D[j]
			if c, ok := countMap[val]; ok {
				res += c
			}
		}
	}
	return res
}
